### PR TITLE
INSP: Detect E0747 error

### PR DESCRIPTION
**Pull request by mchernyavsky**
Relates to https://github.com/intellij-rust/intellij-rust/issues/3985.

changelog: Detect [E0747](https://doc.rust-lang.org/error-index.html#E0747) error

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt [hunk 1]
```diff
@@ -790,8 +800,24 @@ class RsTypeInferenceWalker(
         }
     }
 
-    fun inferConstArgumentTypes(constParameters: List<RsConstParameter>, constArguments: List<RsExpr>) {
-        inferArgumentTypes(constParameters.map { it.typeReference?.type ?: TyUnknown }, constArguments)
+    fun inferConstArgumentTypes(constParameters: List<RsConstParameter>, constArguments: List<RsElement>) {
+        val argDefs = constParameters.asSequence()
+            .map { it.typeReference?.type ?: TyUnknown }
+            .map(ctx::resolveTypeVarsIfPossible)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt:806]: This line is not needed
```suggestion
            
```

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt [hunk 2]
```diff
@@ -654,11 +654,21 @@ class RsTypeInferenceWalker(
...
-            expr.evaluate(expectedTy, resolver)
-        }
-        val constSubst = constParameters.zip(constArguments).toMap()
+        val constSubst = constParameters.zip(methodCall.constArguments).mapNotNull { (param, expr) ->
+            val expectedTy = param.parameter.typeReference?.type ?: TyUnknown
+            param to when (expr) {
+                is RsExpr -> expr.evaluate(expectedTy, resolver)
+                is RsTypeReference -> {
+                    when (val element = (expr as? RsBaseType)?.path?.reference?.resolve()) {
+                        is RsConstant -> element.expr?.evaluate(element.typeReference?.type ?: expectedTy, resolver)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt:663]:
```suggestion
                is RsBaseType -> {
                    when (val element = expr.path?.reference?.resolve()) {
                        is RsConstant -> element.expr?.evaluate(element.typeReference?.type ?: expectedTy, resolver)
```

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt [hunk 3]
```diff
@@ -790,8 +800,24 @@ class RsTypeInferenceWalker(
...
+    fun inferConstArgumentTypes(constParameters: List<RsConstParameter>, constArguments: List<RsElement>) {
+        val argDefs = constParameters.asSequence()
+            .map { it.typeReference?.type ?: TyUnknown }
+            .map(ctx::resolveTypeVarsIfPossible)
+            .infiniteWithTyUnknown()
+        for ((type, expr) in argDefs.zip(constArguments.asSequence())) {
+            when (expr) {
+                is RsExpr -> expr.inferTypeCoercableTo(type)
+                is RsTypeReference -> {
+                    val typeReference = when (val def = (expr as? RsBaseType)?.path?.reference?.resolve()) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt:812]:
```suggestion
                is RsBaseType -> {
                    val typeReference = when (val def = expr.path?.reference?.resolve()) {
```

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt [hunk 4]
```diff
@@ -654,11 +654,21 @@ class RsTypeInferenceWalker(
...
         val constParameters = callee.element.constParameters.map { CtConstParameter(it) }
         val resolver = PathExprResolver.fromContext(ctx)
-        val constArguments = methodCall.constArguments.withIndex().map { (i, expr) ->
-            val expectedTy = constParameters.getOrNull(i)?.parameter?.typeReference?.type ?: TyUnknown
-            expr.evaluate(expectedTy, resolver)
-        }
-        val constSubst = constParameters.zip(constArguments).toMap()
+        val constSubst = constParameters.zip(methodCall.constArguments).mapNotNull { (param, expr) ->
+            val expectedTy = param.parameter.typeReference?.type ?: TyUnknown
+            param to when (expr) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt:659]: This looks like a duplication of code in `RsPsiSubstitution.toSubst`. Is it not possible to deduplicate it somehow?

File: src/main/kotlin/org/rust/ide/inspections/RsWrongGenericArgumentsOrderInspection.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                if (!isPathValid(type.path)) return
+                checkGenericArguments(holder, type)
+            }
+
+            override fun visitTraitRef(trait: RsTraitRef) {
+                if (!isPathValid(trait.path)) return
+                checkGenericArguments(holder, trait)
+            }
+
+            override fun visitCallExpr(element: RsCallExpr) = checkGenericArguments(holder, element)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/RsWrongGenericArgumentsOrderInspection.kt:30]: These 3 options can be unified to `visitPath(RsPath)`. This also takes into account path exprs liks this
```rust
let a = foo::<T>;
a();
```

As I see, this is also actual for `RsWrongGenericArgumentsNumberInspection`

**Comment by mchernyavsky** [comment]
TODO: add tests for E0308

**Comment by vlad20012** [comment]
There are regressions

**Comment by mchernyavsky** [comment]
bors r=vlad20012

**Pull request merged**